feat: multiple accounts support in ledger

dawnseeker8 opened the pull request:
## **Description**
This PR will enable the multiple accounts supports for ledger devices.
Following changes has been made in this PR:
1. use @metamask/eth-ledger-keyring-bridge@4.0.0 to replace old @consensys/ledgerhq-metamask-keyring@0.0.9
2. add `LedgerSelectAccount` component to allow user select multiple accounts from ledger devices. (screen is similiar to QR code select account screen)
3. add `remove accounts` for all hardware wallet accounts in `AccountActions.tsx` file. 
4. add some metric logging for `remove accounts` and `connect accounts`
5. modify the `engine.ts` code and `ledger.ts` to allow intialise the new ledger keyring and its middleware and transport object.
6. Modify the `BlockingActionModel` to support `onAnimationCompleted` event so that we can have better smooth model animation than before. (very lagging animation when some heavy operations like import multiple accounts happened in the background)


<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
7. What is the improvement/solution?
-->

## **Related issues**

Fixes:

## **Manual testing steps**

**Connect Multiple Ledger accounts**

1. Launch wallet
2. Create a new account
3. Once on the wallet view, add a new account or hardware wallet
4. Add hardware wallet
5. Ledger
6. Once device is found, click Continue
7. Tick the selected Ledger accounts to be imported
8. Click Unlock


**Forget Multiple Ledger accounts**

1. Use an existing wallet that has multiple Ledger accounts imported
2. Once on the wallet view, add a new account or hardware wallet
4. Add hardware wallet
5. Ledger
6. Once device is found, click 'Continue'
7. Click 'Forget this device' button
8. User will be returned to account list with all Ledger accounts removed

**Forget Individual Ledger Account**

1. Use an existing wallet that has multiple Ledger accounts imported
2. Once on the wallet view, click on the three dots beside the Address section
3. Click 'Remove hardware account'
4. Click 'Remove'
5. User will be returned to account list with the individual Ledger account removed

**Forget Individual QR Code Wallet Account**

1. Use an existing wallet that has multiple QR Code accounts imported
2. Once on the wallet view, click on the three dots beside the Address section
3. Click 'Remove hardware account'
4. Click 'Remove'
5. User will be returned to account list with the individual QR Code account removed

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

dawnseeker8 committed:
feat: add library @metamask/eth-ledger-bridge-keyring, and the patch the library with local change.
1. modify the Engine.ts to use new library
2. modify Ledger.ts to use new library

will expect some errors during build, will fix this error and unit tests in next commit.

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support

dawnseeker8 committed:
feat: make a backup of the change to switch to another branch.

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support

# Conflicts:
#	ios/MetaMask.xcodeproj/project.pbxproj

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support

# Conflicts:
#	ios/MetaMask.xcodeproj/project.pbxproj

dawnseeker8 committed:
fix: Downgrade the library

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support

# Conflicts:
#	app/core/Engine.ts

dawnseeker8 committed:
fix: Rebase from main

dawnseeker8 committed:
feat: add `package.tgz` to project so that project can build without the need of release of library. also fix some issue of missing deviceId in library cause the bluetooth unable to connect.

dawnseeker8 committed:
feat: revert the ledger.ts code in hardwareWallet folder.

dawnseeker8 committed:
feat: workable version with add multiple accounts and forget devices. Still need to refactory the code to remove unused functions.

dawnseeker8 committed:
feat: workable version with add multiple accounts and forget devices. Still need to refactory the code to remove unused functions.

dawnseeker8 committed:
feat: Fix the unit tests in LedgerConnect

dawnseeker8 committed:
feat: Fix the lint error and redundant imports

dawnseeker8 committed:
feat: Try to fix the hash unmatch issue in @metamask/eth-ledger-keyring

dawnseeker8 committed:
feat: Fix the build issue after the package.tgz file change.

dawnseeker8 committed:
feat: Try to fix the `yarn setup` integrity check failed because checksum not match.
also improve the unit tests for ledger.ts to cover all functions.

dawnseeker8 committed:
feat: make a backup before carry out the rebase.

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support

# Conflicts:
#	app/components/Nav/App/index.js
#	app/core/Engine.ts
#	ios/MetaMask.xcodeproj/project.pbxproj
#	package.json
#	yarn.lock

dawnseeker8 committed:
feat: forget to push `yarn.lock`

dawnseeker8 committed:
feat: Fix the build issue after rebase.

dawnseeker8 committed:
feat: add `remove hardware wallet` from AccountActions.tsx

dawnseeker8 committed:
feat: add `loading` mask when remove hardware account devices.

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support

# Conflicts:
#	app/core/Engine.ts
#	app/core/Ledger/Ledger.test.ts
#	app/core/Ledger/Ledger.ts
#	patches/@metamask+keyring-controller+9.0.0.patch
#	yarn.lock

dawnseeker8 committed:
feat: Rebase from main to upgrade to keyringController 13. Fix all blocken code.

Enhance in BlockingActionModal with smooth animation to provide better user friendly interaction.

add code to remove single accounts in actions.

Still need to fix unit tests, however because MM mobile team will upgrade keyringController to 15, will start to fix unit tests after rerbase.

dawnseeker8 committed:
fix: Backup new metric change.

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support-rebase

# Conflicts:
#	app/components/Views/ConnectHardware/SelectHardware/index.tsx
#	app/components/Views/LedgerAccountInfo/index.tsx
#	app/core/Ledger/Ledger.test.ts
#	app/core/Ledger/Ledger.ts
#	package.json
#	patches/@metamask+keyring-controller+13.0.0.patch
#	yarn.lock

dawnseeker8 committed:
fix: Rebase from main to upgrade to keyring-controller 16.
This version will be first version which ledger work after keyring controller 16 upgrade.

dawnseeker8 committed:
feat: Add unit tests to cover all functions in ledger core.

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support-rebase

dawnseeker8 committed:
feat: Fix the broken unit tests after rebase.

dawnseeker8 committed:
feat: change ho use @metamask/eth-ledger-keyring-bridge@4.0.0 version.

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support-rebase

# Conflicts:
#	app/components/Nav/App/index.js
#	app/components/Views/LedgerConnect/index.tsx
#	app/core/Ledger/Ledger.ts

dawnseeker8 committed:
feat: fix some lint issue after rebase from main

dawnseeker8 committed:
feat: fix some lint issue after rebase from main

dawnseeker8 committed:
feat: fix all unit tests after rebase

dawnseeker8 committed:
feat: fix all lint and tsc in broken pipeline

dawnseeker8 committed:
feat: Fix ide complain about the navigationProp not contains pop method.

Also fix the unit tests broken in Ledger.test.ts

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support-rebase

dawnseeker8 committed:
feat: Remove the redundant file  `LedgerAccountInfo/index.tsx`,

Refine the test description.

dawnseeker8 committed:
feat: Revert the format of file.

dawnseeker8 committed:
feat: revert to main version.

dawnseeker8 committed:
feat: update yarn.lock

dawnseeker8 committed:
feat: update yarn.lock

dawnseeker8 committed:
feat: Remove TODO in the file.

dawnseeker8 committed:
feat: Backup the application to test ios later.

dawnseeker8 committed:
feat: Backup the application to test ios later and remove the TODO comments.

dawnseeker8 committed:
feat: Fix a deviceDiscontinued  issue in ios.

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support-rebase

dawnseeker8 committed:
feat: Fix the build issue after rebase.

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support-rebase

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support-rebase

dawnseeker8 committed:
feat: Improve code coverage for `LedgerSelectAccount` components.

dawnseeker8 committed:
feat: Fix the lint error in test

dawnseeker8 committed:
feat: Fix the lint error in test

dawnseeker8 committed:
feat: code the branch code in `ledger` core

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support-rebase

dawnseeker8 committed:
feat: add unit tests for AccountActions components

dawnseeker8 committed:
feat: Fix the lint issue in unit tests.

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support-rebase

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support-rebase

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support-rebase

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support-rebase

dawnseeker8 commented on app/components/Views/LedgerSelectAccount/index.tsx:
```diff
@@ -0,0 +1,183 @@
+    setAccounts(_accounts);
+  }, []);
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  const onToggle = useCallback(() => {}, []);
```
onToggle function is required in <AccountSelector> component, and the reason i put this method as useCallback to prevent function re-initialization during rerendering even it did nothing.

gantunesr replied:
In that case @dawnseeker8, we can just modify the props interface `ISelectQRAccountsProps` to make it optional and then remove the callback.

It would be good too to rename this interface too to something like `SelectAccountsProps` and the prop name `toggleAccount` to `checkAccount` since this is a check box and not a toggle

dawnseeker8 commented on patches/@metamask+eth-ledger-bridge-keyring+4.0.0.patch:
there is a PR created https://github.com/MetaMask/eth-ledger-bridge-keyring/pull/231 for eth-ledger-bridge-keyring.  
To fix the iOS bug, i need to re-generate the `eth` app when setTransport set.  because `ledger-transport-middleware` was hidden behind the `ledger-mobile-bridge`. there is no way to directly fix this in MM mobile `ledger.ts` file. however, if we can merge above PR and release a new version. we can avoid using this patch.

dawnseeker8 replied:
The patch fix here is slightly difference to PR code fix in eth-ledger-keyring-bridge library after i refine some unit tests.  please follow what https://github.com/MetaMask/eth-ledger-bridge-keyring/pull/231 as final implementation.

dawnseeker8 committed:
Apply suggestions from code review

commit Gustavo Antunes's code suggestion.

Co-authored-by: Gustavo Antunes <17601467+gantunesr@users.noreply.github.com>

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support-rebase

# Conflicts:
#	app/components/Views/AccountActions/AccountActions.constants.ts
#	app/components/Views/AccountActions/AccountActions.test.tsx
#	app/components/Views/AccountActions/AccountActions.tsx
#	package.json
#	yarn.lock

dawnseeker8 committed:
feat: Rebase from main to resolve the conflict.

dawnseeker8 committed:
feat: fix lint error.

dawnseeker8 committed:
feat: fix deduplicate in yarn.lock

dawnseeker8 committed:
feat: Remove the patch to Keyring-controller 16 and move unlockLedgerWalletAccount to Ledger.ts

dawnseeker8 commented on patches/@metamask+keyring-controller+16.0.0.patch:
hi, i have successfully removed all changes in this patch, and move the function `unlockLedgerWalletAccount` to `Ledger.ts` file now.

dawnseeker8 committed:
feat: improve code coverage to Ledger.ts to 100%, remove the whitespace difference in patch file.

dawnseeker8 commented:
Final e2e tests are all green even `Bitrise E2E Status` display failure. https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/c6b59eaa-aacd-4177-acbb-acac8342db20

dawnseeker8 committed:
feat: Change error log title to correct devices.

dawnseeker8 committed:
feat: Fix the data migration issue from ledger accounts from V7.24.4

gantunesr commented on app/components/Views/AccountActions/AccountActions.tsx:
```diff
@@ -164,7 +256,24 @@ const AccountActions = () => {
+        onAnimationCompleted={triggerRemoveHWAccount}
+      >
+        <Text style={styles.text}>
+          {strings('connect_qr_hardware.please_wait')}
```
We should probably change the string identifier from `connect_qr_hardware.please_wait`

dawnseeker8 replied:
yea, let me create a share message key for both QR screen and Ledger screen. i will call it `common.please_wait`

dawnseeker8 committed:
Apply suggestions from code review

Co-authored-by: Gustavo Antunes <17601467+gantunesr@users.noreply.github.com>

dawnseeker8 commented on app/core/Ledger/Ledger.ts:
```diff
@@ -43,46 +46,25 @@ export const connectLedgerHardware = async (
-      // eslint-disable-next-line @typescript-eslint/no-explicit-any
-      keyring.setTransport(transport as unknown as any, deviceId);
-      return await keyring.getAppAndVersion();
+      keyring.setHdPath("m/44'/60'/0'/0");
```
i will create a constant variable file in `Ledger` directory.

dawnseeker8 committed:
feat: commit some change after code review,, Following change has beren done.
1. create a constants.ts under core/ledger directory to contain the LEDGER_HD_PATH
2. add `common.please_wait` message key in en.json, and remove the not common message key for ledger and qr code.

dawnseeker8 committed:
Merge remote-tracking branch 'origin/feat/ledger-multiple-accounts-support-rebase' into feat/ledger-multiple-accounts-support-rebase

dawnseeker8 commented on app/components/Views/AccountActions/AccountActions.tsx:
```diff
@@ -122,6 +137,83 @@ const AccountActions = () => {
+            break;
+          case ExtendedKeyringTypes.qr:
+            await Controller.KeyringController.forgetQRDevice();
+            // there is not a MetaMetricsEvent for this action??
```
I will recreate a new MetricsEvents call `HARDWARE_WALLET_FORGOTTEN`, and put `device_type` to differential them. we can also use it for the future device type.

I will do the simliar MetricsEvents change for `CONNECT_WALLET` , currently there is not `connect QR or Keystone` metric yet, only `CONNECT_LEDGER`

dawnseeker8 committed:
feat: change MetrixEvent for HARDWARE_WALLET_FORGOTTEN with deviceType property.

dawnseeker8 commented on app/components/Views/AccountActions/AccountActions.tsx:
```diff
@@ -122,6 +137,83 @@ const AccountActions = () => {
+    );
+  }, []);
+
+  const triggerRemoveHWAccount = useCallback(async () => {
```
i will do a refactor of this function to split the function to smaller pieces.

dawnseeker8 committed:
feat: Refactor the `getLedgerAccountsByPage`to `getLedgerAccountsByOperation`. and put opeartionType enuim to constants.ts

dawnseeker8 committed:
feat: fix the format issue in the file.

dawnseeker8 committed:
feat: refactor to put OperationTypes to enum in `types.ts` file.

dawnseeker8 committed:
feat: Create HardwareDeviceTypes enum to MetaMetrics.types.ts, and refactor all hardware relative metricEvent to use that new enum to keep all deviceTypes consistency.

dawnseeker8 committed:
feat: refactor `triggerRemoveHWAccount` method to smaller pieces, introduce `removeHardwareAccount`, `reselectFirstAccount` and `forgetDeviceIfRequired`

dawnseeker8 commented on app/core/Ledger/Ledger.ts:
```diff
@@ -43,46 +46,25 @@ export const connectLedgerHardware = async (
+      keyring.setHdPath("m/44'/60'/0'/0");
+      keyring.setDeviceId(deviceId);
+
+      const bridge = keyring.bridge as LedgerMobileBridge;
```
`LedgerMobileBridge` has three extra functions which `LedgerBridge` hasn't included.  the default Keyring.bridge is `LedgerBridge` type. that is why i need to use `as` to let typescript compiler know that bridge here is `LedgerMobileBridge`.

However, according to your suggestion, i have made a following [PR](https://github.com/MetaMask/eth-ledger-bridge-keyring/pull/233) to redefine the type hierarchy. which LedgerMobileBridge extends LedgerBridge. 

Let me know what you think about this change.

gantunesr replied:
LGTM, I have merged that PR

gantunesr commented on app/components/Views/LedgerSelectAccount/index.tsx:
```diff
@@ -0,0 +1,193 @@
+          await unlockLedgerWalletAccount(index);
+        }
+      } catch (err) {
+        Logger.log('Error: Connecting Ledger hardware wallet', err);
```
Lets omit this, I don't think it's relevant to log it since the connection error could be that the device is not unlocked

```suggestion
        // Do nothing
```

gantunesr commented on app/components/Views/LedgerSelectAccount/index.tsx:
```diff
@@ -0,0 +1,193 @@
+
+    if (forgetDevice) {
+      await onForget();
+      // ...
```
Were you going to add a comment here?

dawnseeker8 replied:
Weird, i have no idea when this // ... has been added, probably generate from refactor of the code. i will remove the comment here.

gantunesr commented on app/core/Analytics/MetaMetrics.events.ts:
```diff
@@ -353,7 +353,9 @@ enum EVENT_NAME {
-  LEDGER_HARDWARE_WALLET_FORGOTTEN = 'Ledger hardware wallet forgotten',
+
+  // common hardware wallet
+  HARDWARE_WALLET_FORGOTTEN = 'Hardware wallet forgotten',
```
Did @AlexJupiter approved this event name change? This could cause bugs in our analytics dashboards

dawnseeker8 replied:
i will revert this change at the moment, and i have already posted a message to @AlexJupiter to confirm this change. if answer is yes, i will re-introduce this change. 

if answer is no, QR_HARDWARE_WALLET_FORGOTTEN may need to be introduced for `QR hardware wallet forgooten` event.

dawnseeker8 replied:
@AlexJupiter has agreed to use `HARDWARE_WALLET_FORGOTTEN` with device_type for all hardware relative forgotten metric event.

gantunesr commented on app/core/Analytics/MetaMetrics.types.ts:
```diff
@@ -152,3 +152,8 @@ export interface IDeleteRegulationStatus {
   hasCollectedDataSinceDeletionRequest: boolean;
   dataDeletionRequestStatus: DataDeleteStatus;
 }
+
+export enum HardwareDeviceTypes {
+  LEDGER = 'Ledger',
+  QR = 'QR Hardware',
+}
```
In other parts of the code you're using `ExtendedKeyringTypes`, I think it's better to just use this for analytics too to maintain consistency across the codebase

dawnseeker8 replied:
i will move the this `enum HardwareDeviceTypes` to where `ExtendedKeyringTypes` is.

gantunesr commented on app/core/Ledger/Ledger.ts:
```diff
@@ -110,7 +95,41 @@ export const forgetLedger = async (): Promise<void> => {
+      }
+    });
+
+    console.warn(`Returning accounts by ${operation}`, accounts);
```
Lets remove this

dawnseeker8 replied:
Sorry, my mistaken during paging issue debugging.

gantunesr reviewed (commented):
Last few comments! @dawnseeker8 can you also add a recording of the feature?

dawnseeker8 committed:
feat: remove console log.

dawnseeker8 committed:
Apply suggestions from code review

Co-authored-by: Gustavo Antunes <17601467+gantunesr@users.noreply.github.com>

dawnseeker8 commented on app/components/Views/AccountActions/AccountActions.tsx:
```diff
@@ -122,6 +139,122 @@ const AccountActions = () => {
+   * @param keyring - The keyring object
+   */
+  const forgetDeviceIfRequired = useCallback(
+    async (keyring: KeyringObject) => {
```
yea, you are right, i have change the method argument to `keyringType: string` here.

dawnseeker8 committed:
feat: refactor the code according to Gustova's code review comments

dawnseeker8 committed:
feat: Remove cancel button onPress event and also remove corresponding unit test for cancel button click.

dawnseeker8 commented on app/components/Views/AccountActions/AccountActions.tsx:
```diff
@@ -122,6 +139,122 @@ const AccountActions = () => {
+      [
+        {
+          text: strings('accounts.remove_account_alert_cancel_btn'),
+          onPress: () => false,
```
I try to remove it, but unit test for click `cancel` button start to failed due to onPress() has been removed.
anyway, i have moved `onPress` event here for react native alert. and then remove the corresponding unit test in test file.

dawnseeker8 committed:
feat: Revert the metric change for `HARDWARE_WALLET_FORGOTTEN` event.

dawnseeker8 committed:
feat: Change to use `selectSelectedInternalAccount` to get the selectedAccount which contain selectedAddress, and metadata

dawnseeker8 commented on app/components/Views/AccountActions/AccountActions.tsx:
```diff
@@ -122,6 +139,122 @@ const AccountActions = () => {
+        return;
+      }
+
+      await removeHardwareAccount(keyring, selectedAddress as Hex);
```
have changed to use `selectSelectedInternalAccount` and remove all argument for removeHardwareAccount to use selectedAddress as selectedAccount.address and keyring = selectedAccount.metadata.keyring.

dawnseeker8 committed:
feat: Alex allow to use Hardware_wallet_forgotten metric event.

dawnseeker8 committed:
feat: fix the lint issue

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support-rebase

# Conflicts:
#	yarn.lock

dawnseeker8 committed:
feat: fix the yarn deduplicate

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support-rebase

dawnseeker8 committed:
feat: upgrade the eth-ledger-keyring-bridge library, remove the patch for that library

dawnseeker8 committed:
feat: to make account selection screen bigger so that more items can be shown in the screen.

dawnseeker8 commented on app/components/UI/HardwareWallet/AccountSelector/styles.tsx:
```diff
@@ -50,7 +50,7 @@ export const createStyle = (colors: any) =>
       alignItems: 'center',
       justifyContent: 'space-between',
-      paddingTop: 70,
+      paddingTop: 30,
```
This styling change is to make the account selection area visually bigger in the small device screen so that user can see five accounts in the screen.

dawnseeker8 committed:
feat: Fix the remove account not log metric in mixpanel issue.

gantunesr commented on app/components/Views/LedgerSelectAccount/index.tsx:
```diff
@@ -0,0 +1,191 @@
+  }, []);
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  const onToggle = useCallback(() => {}, []);
```
Regarding the previous comment @dawnseeker8, we can just modify the props interface `ISelectQRAccountsProps` to make it optional and then remove the callback.

It would be good too to rename this interface to something like `SelectAccountsProps` and the prop name `toggleAccount` to `onCheck` since this is a check box and not a toggle

dawnseeker8 replied:
I have changed the `AccountSelector` component to make `onCheck` (onToggle) property optional, and change `ConnectQRHardware` and `LedgerSelectAccount` component code to reflect above property change.

gantunesr reviewed (commented):
LGTM! I just left one final comment that was not resolved yet in the previous reviews

dawnseeker8 committed:
feat: Modify the AccountSelector component to make `onCheck` optional, and then modify `LedgerSelectAccount` component and `ConnectQRHardware` component to reflect the property changed.

dawnseeker8 commented:
> LGTM! I just left one final comment that was not resolved yet in the previous reviews

hi, @gantunesr i have done the change to `AccountSelector` component to make `onCheck` (toggleAccount) optional.

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-multiple-accounts-support-rebase

vivek-consensys commented:
Flows tested on iPhone 13/iOS 17.5.1 and Samsung S21/Android 14 and screen recordings attached below:-

**Connect Multiple Ledger accounts**

https://github.com/user-attachments/assets/11c73aab-60f3-44ca-a56c-f8e62949e47d

**Forget Multiple Ledger accounts**

[Private Zenhub Video](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBODNnQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--75c12a1fc9b8294a2e2c5c347c81731053b35951/Scenario2_iOS.MP4)


**Forget Individual Ledger Account**

**Forget Individual QR Code Wallet Account**

dawnseeker8 commented:
Final `e2e` tests has passed
https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/942d85f4-ec3c-4bab-936f-333db0f572cd

dawnseeker8 committed:
feat: Change the structure of `LedgerSelectAccount` and `LedgerConnect` component to include a close icon, which will minic how the `ConnectQRHardware` component structure. hide the navigation header for `ledger Connect` screen.

tommasini reviewed (approved):
LGTM!

dawnseeker8 commented:
E2e smoke tests pass. https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/92c7ec10-e428-40d5-b64c-41cc42c0c577
